### PR TITLE
Update production deploy sop

### PR DESCRIPTION
**Pull request by kencho51**
# Pull request for issue: #1421

This is a pull request for the following functionalities:

- Updated the terraform and ansible usages in `PRODUCTION_DEPLY.md`
- Start to curate a list of problem/solution set in the `PRODUCTION_TROUBLESHOOT.md`, which will keep growing to align with the application growth

## Prerequisite

The DNS A record of `gigadb.org` has been updated which it points to the elastic IP of the production webapp server.

## How to test?

Go through the steps in the updated SOP  `PRODUCTION_DEPLOY.md` again.
If encounter problems, try to seek help from the `PRODUCTION_TROUBLE.md`.

## How have functionalities been implemented?

None.

## Any issues with implementation?

None.

## Any changes to automated tests?

None.

## Any changes to documentation?

- Renamed the `BETA_DEPLOY.md` to `PRODUCTION_DEPLY.md`
- Created `AWS_SETUP.md`  for one off  `AWS IAM policies` and `Set up credentials for accessing AWS resources` steps.
- Created `PRODUCTION_TROUBLE.md` to curate possible solution/fix for the problems during the deployment process.

## Any technical debt repayment?

Created `.gitignore` in `ops/infrastructure/envs`, so the staging dir and live dir would not be tracked, as a  result, the whole staging dir or live could be deleted if a fresh provisioning is needed. 

## Any improvements to CI/CD pipeline?

None.

### CODE REVIEW

**Review by rija — CHANGES_REQUESTED** [review]
Hi @kencho51,

**praise**: documentation is thorough and is actionable (tried on Upstream's staging)

**issue**: that doc is still very dense and long, which doesn't work as a tool to be used in operational context
**suggestion**: Move (just leave a link) the "Set up credentials for accessing AWS resources" section into its own doc (AWS_SETUP.md) because:

		1) Its content is only done once (whereas the actions in the remainder of the doc is done every time we rebuild environments), 
		2) there is more than one way to do it (e.g: the way I do for myself is not the way described in the doc)
 **suggestion**: for the same reason as above, move (and leave a link) the "AWS IAM policies" section elsewhere (maybe in the same AWS_SETUP.md doc mentioned above)
**suggestion**: Don't put an extract of the terraform state since we can download it. However, instead of linking to the word "here" (twice), put the link on description of what is it linked. Eg.: "Shared Terraform state for Upstream's staging environment" for the first link, "Shared Terraform state for Upstream's live environment" for the second link
**suggestion**: remove mention of "beta", this is now the real website, also rename the doc to `PRODUCTION_DEPLOY.md`

**issue**: given we can retrieve the state from Gitlab, the local directories `envs/staging` or `envs/live` are ephemeral and disposable but the codebase still treat them as git versioned and expects a `.gitignore` inside, causing all the provisioning artefacts to appear as new files in your git tree after you've recreated these directories
**suggestion**: replace `envs/(staging|live)/.gitignore` files with a single `envs/.gitignore` (and adjusts the paths inside accordingly), and remove `envs/(staging|live)` directory from git

**issue**: the "Deploy to staging.gigadb.org using CI/CD pipeline" section doesn't mention Tideways deployment
**suggestion**: Add instructions to not forget to trigger the build and deployment  of Tideways to staging and live (also related to #1444)

**issue**: as an operational doc, it's lacking what to do in case of problems
**suggestion**: Link to a  "PRODUCTION_TROUBLESHOOT.md" doc that has these headings (and corresponding solutions) (and add more if you think of something else)
	* How to renew an expiring TLS certificate
	* What to do if we receive disk usage alerts or pipeline jobs fail due lack of space on device
	* How to deploy a previous version of the software if the current deployment has major bugs
	* What to do if terraform execution fails
	* What to do if Ansible playbook execution fails (related to #1443)
	* What if Ansible playbook cannot find host IP addresses

File: docs/sop/PRODUCTION_DEPLOY.md [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+4. Once logged into the console, go to the Domain Names page
+5. You will see an entry for `gigadb.org` domain - click on this
+6. You will now see the `DNS Settings gigadb.org` page. There should be an entry
+for the Host with a value equal to the `eip-gigadb-live-gigadb` elastic 
+IP address. There will also be an entry for the `staging` Host too.
+
+### Set up credentials for accessing AWS resources
+[Click to see how to setup credentials](AWS_SETUP.md#set-up-credentials-for-accessing-aws-resources)
+
+### Tools
```
Comment (pli888) [inline docs/sop/PRODUCTION_DEPLOY.md:53]: **thought**: As I found out from trying to access AWS resources in a new Gigadb user account last week, other tools like Docker Desktop, Terraform, Ansible, terraform-inventory, AWS CLI, jq are required too. We take for granted that these tools have already been installed from our developer work.

File: docs/sop/PRODUCTION_DEPLOY.md [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+```
+
+Provision monitoring server:
+```
+$ env OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES ansible-playbook -i ../../inventories monitoring_playbook.yml
+```
+
+Additional features for executing ansible playbooks:
+```
+# display all availale plays
```
Comment (pli888) [inline docs/sop/PRODUCTION_DEPLOY.md]: Typo in `availale`.

File: docs/sop/PRODUCTION_DEPLOY.md [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
```
Comment (pli888) [inline docs/sop/PRODUCTION_DEPLOY.md]: Typo in `availale`.

File: docs/sop/PRODUCTION_TROUBLESHOOT.md [hunk 4]
```diff
@@ -0,0 +1,275 @@
+# Troubleshooting guide for the GigaDB Website deployment
```
Comment (pli888) [inline docs/sop/PRODUCTION_TROUBLESHOOT.md:1]: **question**: Are we still using portainer? If so, can instructions for accessing it be added into this document?

**Review by pli888 — COMMENTED** [review]
**praise**: The documentation looks good to me.
**suggestion**: I've made a couple of suggestions in the code.
**chore**: Rebase with latest develop code and resolve conflict in changelog file.

File: docs/sop/PRODUCTION_DEPLOY.md (hunk 1 above)
Comment (rija) [reply docs/sop/PRODUCTION_DEPLOY.md:53]: @pli888, @kencho51 

**suggestion**: ensure jq, terraform-inventory, ansible are listed in the Tools section of `PRODUCTION_DEPLOY.md`'s prerequisite section
**suggestion**: AWS CLI should be mentioned in the `AWS_SETUP.md`, not here

File: docs/sop/PRODUCTION_DEPLOY.md [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+$ git log
+```
+
+If not, then get the latest code:
+```
+$ git fetch origin
+$ git rebase origin/develop
+```
+
+Check you can deploy GigaDB in your `dev` environment:
```
Comment (rija) [inline docs/sop/PRODUCTION_DEPLOY.md]: **issue (non-blocking)**: I feel like L 108-122 are unnecessary in `PRODUCTION_DEPLOY.md`, those instructions are already in `README.md`
**suggestion**: Maybe add a line to tell the operator to ensure it has deployed the `develop` branch in its own AWS deployment.

File: docs/sop/PRODUCTION_DEPLOY.md [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+
+### Prerequisites
+
+You need to have permission to push or merge to the protected `develop` branch
+in the `Upstream` group in order to run manual jobs such as `build_staging` in 
+the CI/CD pipeline otherwise you will see a `You are not authorised to run this
+manual job` message. To see who can push/merge, go to repository settings and 
+expand the Protected branches section for the gigadb-website project.
+
+### Procedure
```
Comment (rija) [inline docs/sop/PRODUCTION_DEPLOY.md:232]: **issue**: On staging, build and deploy are automated. 
**suggestion**: That should be mentioned here. And then, just add a comment that the build and deploy jobs can be rerun if necessary.

File: docs/sop/PRODUCTION_DEPLOY.md [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+And you have to make sure that terraform state files exist in [gitlab terraform page](https://gitlab.com/gigascience/upstream/gigadb-website/-/terraform).
+You can click to download the [shared Terraform state for Upstream's staging environment](https://gitlab.com/api/v4/projects/11385199/terraform/state/staging_infra/versions/144)
+for checking in the latter staging deployment steps
+
+Change directory to the `envs` folder:
+```
+$ cd ops/infrastructure/envs
+```
+
+Change directory to the `staging` folder:
```
Comment (rija) [inline docs/sop/PRODUCTION_DEPLOY.md]: **issue**: that directory (and the `live` directory in the next section) are no longer part of the codebase
**suggestion**: the sentence  on L149 should read with something like "Create directory `staging` if not existing  already and then change to it.

File: docs/sop/PRODUCTION_DEPLOY.md [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+
+And also, you can click to download the [shared Terraform state for Upstream's live environment](https://gitlab.com/api/v4/projects/11385199/terraform/state/live_infra/versions/76)
+for checking in the latter live deployment steps.
+
+Change directory to the `envs` folder:
+```
+$ cd ops/infrastructure/envs
+```
+
+Change directory to the `live` directory:
```
Comment (rija) [inline docs/sop/PRODUCTION_DEPLOY.md]: **issue**: see my comment above about the `staging` directory as `live` has same issue

File: docs/sop/PRODUCTION_DEPLOY.md [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+Copy terraform files to `staging` environment:
+```
+# this step will retrieve terraform state file from the gitlab
+$ ../../../scripts/tf_init.sh --project gigascience/upstream/gigadb-website --env staging
+
+You need to specify the path to the ssh private key to use to connect to the EC2 instance: 
+~/.ssh/id-rsa-aws-hk-gigadb.pem
+
+You need to specify your GitLab username:
+pli888 | rija | kencho51
```
Comment (rija) [inline docs/sop/PRODUCTION_DEPLOY.md]: **issue**: no need to list usernames, it is security issue (leaking system info)
**suggestion**: remove usernames

File: docs/sop/PRODUCTION_DEPLOY.md [hunk 10]
```diff
@@ -0,0 +1,344 @@
...
+$ ../../../scripts/tf_init.sh --project gigascience/upstream/gigadb-website --env live
+
+You need to specify an AWS region:
+ap-east-1
+
+You need to specify the path to the ssh private key to use to connect to the EC2 instance: 
+~/.ssh/id-rsa-aws-hk-gigadb.pem
+
+You need to specify your GitLab username:
+pli888 | rijam | kencho51
```
Comment (rija) [inline docs/sop/PRODUCTION_DEPLOY.md]: **issue**: see my comment above about usernames

File: docs/sop/PRODUCTION_TROUBLESHOOT.md (hunk 4 above)
Comment (rija) [reply docs/sop/PRODUCTION_TROUBLESHOOT.md:1]: @pli888, @kencho51 

**suggestion**: Yes, we are using Portainer, and the instruction is a URL and login/password saved in GitLab variables. And I agree it should be mentioned here

* https://portainer.staging.gigadb.org/
* https://portainer.beta.gigadb.org/

File: docs/sop/PRODUCTION_DEPLOY.md [hunk 11]
```diff
@@ -0,0 +1,344 @@
...
+```
+
+Install third party Ansible roles:
+```
+$ ansible-galaxy install -r ../../../infrastructure/requirements.yml
+```
+
+Provision RDS via bastion server:
+```
+$ env OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES ansible-playbook -i ../../inventories bastion_playbook.yml --extra-vars="gigadb_env=staging"
```
Comment (rija) [inline docs/sop/PRODUCTION_DEPLOY.md]: **typo**: we are in the section for provisioning live, but  I see `--extra-vars="gigadb_env=staging"` here
**suggestion**: replace with `--extra-vars="gigadb_env=live"`

File: docs/sop/PRODUCTION_DEPLOY.md [hunk 12]
```diff
@@ -0,0 +1,344 @@
...
+```
+
+Enable cronjob for periodically resetting the database:
+```
+$ env OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES ansible-playbook -i ../../inventories bastion_playbook.yml -e "reset_database_cronjob_state=present" --extra-vars="gigadb_env=staging"
```
Comment (rija) [inline docs/sop/PRODUCTION_DEPLOY.md]: **typo**: we are in the section for provisioning live, but  I see `--extra-vars="gigadb_env=staging"` here
**suggestion**: replace with `--extra-vars="gigadb_env=live"`

File: docs/sop/PRODUCTION_DEPLOY.md [hunk 13]
```diff
@@ -0,0 +1,344 @@
...
+```
+
+Enable cronjob for periodically creating a database dump and storing it in S3:
+```
+$ env OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES ansible-playbook -i ../../inventories bastion_playbook.yml -e "upload_database_backup_to_S3_cronjob_state=present" --extra-vars="gigadb_env=staging"
```
Comment (rija) [inline docs/sop/PRODUCTION_DEPLOY.md]: **typo**: we are in the section for provisioning live, but  I see `--extra-vars="gigadb_env=staging"` here
**suggestion**: replace with `--extra-vars="gigadb_env=live"`

File: docs/sop/PRODUCTION_DEPLOY.md [hunk 14]
```diff
@@ -0,0 +1,344 @@
...
+# Get public ip of bastion server
+$ terraform output
+
+# ssh into it
+$ ssh -i ~/.ssh/id-rsa-aws-hk-gigadb.pem centos@<bastion public ip>
+```
+
+Provision web application server:
+```
+$ TF_KEY_NAME=private_ip OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES ansible-playbook -i ../../inventories webapp_playbook.yml --extra-vars="gigadb_env=staging"
```
Comment (rija) [inline docs/sop/PRODUCTION_DEPLOY.md]: **typo**: we are in the section for provisioning live, but  I see `--extra-vars="gigadb_env=staging"` here
**suggestion**: replace with `--extra-vars="gigadb_env=live"`

**Review by rija — CHANGES_REQUESTED** [review]
Hi @kencho51 

praise: having separate docs for AWS setup is welcome
praise: having separate docs for troubleshooting is welcome
praise: nice (an useful) touch having incorporated the tips about using `--tags` and `--skip-tags` in the docs 
praise: great having removed wordings related to "beta"
chore: remember to rebase

please also have a look at the inline comments for issues.

File: docs/sop/PRODUCTION_TROUBLESHOOT.md (hunk 4 above)
Comment (kencho51) [reply docs/sop/PRODUCTION_TROUBLESHOOT.md:1]: Hi,

After the live deployment, the live url will be updated to `gigadb.org`, so the portainer url should be `https://portainer.gigadb.org` which I would use in the troubleshooting guide.

File: docs/sop/PRODUCTION_DEPLOY.md (hunk 5 above)
Comment (kencho51) [reply docs/sop/PRODUCTION_DEPLOY.md]: Hi @rija,

This step is to ensure the upstream gigadb website `develop` branch can be spun up  in user's dev environment or not, therefore I would update L108 to 'Check you can spin up GigaDB `develop` branch in your `dev` environment:'

**Review by rija — COMMENTED** [review]
Hi @kencho51 ,
**praise**: all the blocking issues were addressed
**issue**: there should not be a mention of `monitoring_playbook.yml` in `PRODUCTION_DEPLOY.md` (it is present twice, for staging and live instructions), because the grafana/prometheus server is setup once to monitor ALL environments, so it's not part of the provisioning/deployment routine
**chore**: rebase as there is conflict with CHANGELOG.md

**Review by rija — APPROVED** [review]
Hi @kencho51, 
LGTM, happy to approve

**Review by pli888 — APPROVED** [review]
**praise**: The documentation looks good to me after the changes.
**chore**: Need to rebase as there seems to be a conflict in changelog file.

**Pull request merged**